server/eventlog.go

```diff
@@ -422,6 +458,107 @@ func getGrubState(hash crypto.Hash, events []*pb.Event) (*pb.GrubState, error) {
 	return &pb.GrubState{Files: files, Commands: commands}, nil
 }
 
+func getEfiState(hash crypto.Hash, events []*pb.Event) (*pb.EfiState, error) {
+	// We pre-compute various event digests, and check if those event type have
+	// been modified. We only trust events that come before the
+	// ExitBootServices() request.
+	separatorInfo := getSeparatorInfo(hash)
+
+	hasher := hash.New()
+	hasher.Write([]byte(CallingEFIApplication))
+	callingEFIAppDigest := hasher.Sum(nil)
+
+	hasher.Reset()
+	hasher.Write([]byte(ExitBootServicesInvocation))
+	exitBootSvcDigest := hasher.Sum(nil)
+
+	var efiAppStates []*pb.EfiApp
+	var seenSeparator4 bool
+	var seenSeparator5 bool
+	var seenCallingEfiApp bool
+	var seenExitBootServices bool
+	for _, event := range events {
+		index := event.GetPcrIndex()
+		if index != 4 && index != 5 {
+			continue
+		}
+		evtType := event.GetUntrustedType()
+
+		switch index {
+		case 4:
+			// Process Calling EFI Application event.
+			if bytes.Equal(callingEFIAppDigest, event.GetDigest()) {
+				if evtType != EFIAction {
+					return nil, fmt.Errorf("PCR%d contains CallingEFIApp event but non EFIAction type: %d",
+						index, event.GetUntrustedType())
+				}
+				if !event.GetDigestVerified() {
+					return nil, fmt.Errorf("unverified CallingEFIApp digest for PCR%d", index)
+				}
+				// We don't support calling more than one boot device.
+				if seenCallingEfiApp {
+					return nil, fmt.Errorf("found duplicate CallingEFIApp event in PCR%d", index)
+				}
+				if seenSeparator4 {
+					return nil, fmt.Errorf("found CallingEFIApp event in PCR%d after separator event", index)
+				}
+				seenCallingEfiApp = true
+			}
+
+			if evtType == EFIBootServicesApplication {
```

Comment:
Before appending the efi hash, seems like need to check `seenCallingEfiApp` equals true here

Reply:
Nice catch, thanks!